File: 104.Maximum-Depth-of-Binary-Tree.go
// https://leetcode.com/problems/maximum-depth-of-binary-tree/
//
// algorithms
// Easy (58.19%)
// Total Accepted:    422,775
// Total Submissions: 726,507

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var res int

func maxDepth(root *TreeNode) int {
	res = -1

	resursive(root, 0)

	return res
}

func resursive(root *TreeNode, height int) {
	if root == nil {
		res = max(res, height)
		return
	}
	resursive(root.Left, height+1)
	resursive(root.Right, height+1)
}

func max(a, b int) int {
	if a >= b {
		return a
	}
	return b
}

File: 104.Maximum-Depth-of-Binary-Tree.java
// https://leetcode.com/problems/maximum-depth-of-binary-tree/
//
// algorithms
// Easy (60.50%)
// Total Accepted: 507,058
// Total Submissions: 838,100
// beats 100.0% of java submissions


/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {
    public static int depth;

    public int maxDepth(TreeNode root) {
        Solution.depth = Integer.MIN_VALUE;

        recursive(root, 0);

        return Solution.depth;
    }

    public void recursive(TreeNode root, int curDepth) {
        if (root == null) {
            Solution.depth = Math.max(Solution.depth, curDepth);
            return;
        }

        recursive(root.left, curDepth + 1);
        recursive(root.right, curDepth + 1);
    }
}